mithril-common/src/test_utils/certificate_chain_builder.rs

```diff
@@ -419,29 +419,58 @@ impl<'a> CertificateChainBuilder<'a> {
         certificate
     }
 
+    fn update_certificate_previous_hash(
+        &self,
+        certificate: Certificate,
+        previous_certificate: Option<&Certificate>,
+    ) -> Certificate {
+        let mut certificate = certificate;
+        certificate.previous_hash = previous_certificate
+            .map(|c| c.hash.to_string())
+            .unwrap_or_default();
+        certificate.hash = certificate.compute_hash();
+
+        certificate
+    }
+
+    fn fetch_previous_certificate_from_chain<'b>(
+        &self,
+        certificate: &Certificate,
+        certificates_chained: &'b [Certificate],
+    ) -> Option<&'b Certificate> {
+        let is_certificate_first_of_epoch = certificates_chained
+            .last()
+            .map(|c| c.epoch != certificate.epoch)
+            .unwrap_or(true);
+
+        certificates_chained
+            .iter()
+            .rev()
+            .filter(|c| {
+                if is_certificate_first_of_epoch {
+                    // The previous certificate of the first certificate of an epoch
+                    // is the first certificate of the previous epoch
+                    c.epoch == certificate.epoch.previous().unwrap()
+                } else {
+                    // The previous certificate of not the first certificate of an epoch
+                    // is the first certificate of the epoch
```

Comment:
It didn’t seem clear to me, do you think this is better?
```suggestion
                    // The previous certificate that is not the first certificate of an epoch
                    // is the first certificate of the current epoch
```